Testing the GC scanning of flattened value types in arrays

XxAdi101xX opened the pull request:
This commit adds tests for the GC scanning of flattened arrays.

Signed-off-by: Adithya Venkatarao <adi_101@live.com>

XxAdi101xX commented:
@tajila tests have passed on x86-64_linux and x86-64_windows, please review

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1524,6 +1529,133 @@ static public void testCheckCastRefClassOnNull() throws Throwable {
+	//  * single time. This forces the GC to unload the classes.
+	//  */
+	// @Test(priority=4)
+	// static public void testPersistantValueWithSingleAlignment() throws Throwable {
```
Please rename to `testPersistantValueWithSingleAlignment()` to `testValueWithSingleAlignmentGCScanning()` same for the rest

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1524,6 +1529,133 @@ static public void testCheckCastRefClassOnNull() throws Throwable {
+	//  */
+	// @Test(priority=4)
+	// static public void testPersistantValueWithSingleAlignment() throws Throwable {
+	// 	String fields[] = {
```
instead of repeating this in all the similar tests, you can make this a global static `typeWithSingleAlignmentFields`

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1524,6 +1529,133 @@ static public void testCheckCastRefClassOnNull() throws Throwable {
+	// 		"i:QValueInt;:value",
+	// 		"f:QValueFloat;:value",
+	// 		"tri2:QTriangle2D;:value"};
+	// 	Class assortedValueWithSingleAlignmentClass = ValueTypeGenerator.generateValueClass("ValueTypeWithSingleAlignment", fields);
```
Also, this class is being defined elsewhere, so you can turn that variable into a global static

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1524,6 +1529,133 @@ static public void testCheckCastRefClassOnNull() throws Throwable {
+	// 		"tri2:QTriangle2D;:value"};
+	// 	Class assortedValueWithSingleAlignmentClass = ValueTypeGenerator.generateValueClass("ValueTypeWithSingleAlignment", fields);
+
+	// 	MethodHandle makeAssortedValueWithSingleAlignment = lookup.findStatic(assortedValueWithSingleAlignmentClass,
```
likewise with this

tajila commented on runtime/bcutil/ClassFileOracle.cpp:
```diff
@@ -1484,11 +1484,21 @@ ClassFileOracle::walkMethodCodeAttributeCode(U_16 methodIndex)
+			 * state for value types are not in sync with what the stack map expects so we disable it altogether
+			 */
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			if ('Q' == methodDescriptorData[codeIndex]) {
```
`methodDescriptorData[codeIndex]` this isnt actually correct, youre getting lucky because the aload0 is the first bytecode. The methodDesc is essentially the signature for the method, it has nothing to do with the btyecode index.

We will need to do a bigger investigation for this. Continue to do your testing with this code, but remove it from the PR once everything is ready. Remember to also run the internal value type PR build

XxAdi101xX commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1315,15 +1299,7 @@ static public void testDefaultValues() throws Throwable {
-				"d:QValueDouble;:value",
-				"i:QValueInt;:value",
-				"tri:QTriangle2D;:value"};
+		/* Test with assorted ref object with long alignment */
```
The fields removed in this method were all unused.

XxAdi101xX commented:
@tajila I've addressed all the comments, the commit does pass the internal valhalla tests build for with flattening disabled when the tests were uncommented, please take a look at it again

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1524,6 +1498,106 @@ static public void testCheckCastRefClassOnNull() throws Throwable {
+	//  */
+	// @Test(priority=4)
+	// static public void testValueWithLongAlignmentGCScanning() throws Throwable {
+	// 	assortedValueWithLongAlignmentClass = ValueTypeGenerator.generateValueClass("ValueTypeWithLongAlignment", typeWithLongAlignmentFields);
```
the classes dont need to be re-created, same with the constructor MH

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1524,6 +1498,106 @@ static public void testCheckCastRefClassOnNull() throws Throwable {
+	//  * Maintain a buffer of flattened arrays with long-aligned valuetypes while keeping a certain amount of classes alive at any 
+	//  * single time. This forces the GC to unload the classes.
+	//  */
+	// @Test(priority=4)
```
set the priority to 5 so it runs after the tests that create classes have completed

tajila commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1524,6 +1498,106 @@ static public void testCheckCastRefClassOnNull() throws Throwable {
 	}
 
+
+	// The three following tests can be used to verify that flattened value types in arrays are handled properly by the GC. In the
```
Add -Xint to the playlist.xml java command. There is no JIT support for any of the  VT bytecodes right now.

XxAdi101xX commented:
@tajila addressed comments regarding variables

tajila commented:
@gacholio Please review these changes

tajila commented:
@youngar We are adding more tests that will trigger GC scanning of flattened arrays. These tests are commented out, but you can uncomment them for your purposes.

gacholio commented:
This appears to contain real code (not just test code) with no justification that I can see.

tajila commented:
> This appears to contain real code (not just test code) with no justification that I can see.

Yes `hostClass;` is missing from `J9ArrayClass`, this means that `((J9ArrayClass *)clazz)->flattenedElementSize` returns the wrong thing. This causes issues with DDR code.

Would you prefer that part to be put in a separate PR?

gacholio commented:
If the test code is unrelated to that fix, it should probably be separate.

tajila commented:
@XxAdi101xX please separate the VM fix from this PR and place it in a new one

gacholio commented:
Is the VM fix required for these tests to pass?

XxAdi101xX commented:
No they're not. Plus the tests that I've added are commented out anyways and the other changes are minor ones that don't use the flattened field type.

XxAdi101xX commented:
@gacholio I've removed the VM change

gacholio commented:
You should probably get someone from test to review these changes (I'll take a glance, but I doubt I'll have any useful comments).

llxia commented on test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java:
```diff
@@ -1524,6 +1497,87 @@ static public void testCheckCastRefClassOnNull() throws Throwable {
+	//  * Maintain a buffer of flattened arrays with long-aligned valuetypes while keeping a certain amount of classes alive at any 
+	//  * single time. This forces the GC to unload the classes.
+	//  */
+	// @Test(priority=5)
```
I believe you can use `@Test(enabled=false)` to disable the test.

gacholio commented:
I'm not going to merge this - @smlambert can you please have someone look at this?

smlambert commented:
Looks like @llxia is already looking at it based on her note above... assume the only reason to comment out versus disable with annotation would be if this test code does not compile without some feature/code being available yet, but see that Adi has pushed an update, so I can fire off some PR testing

smlambert commented:
Jenkins test sanity all jdk8,jdk11

XxAdi101xX commented:
I'd also run valhalla pipeline with these changes and the tests are passing, not sure if @llxia wants to check off on anything else but currently these changes seem safe.

smlambert commented:
Thanks @XxAdi101xX !

This PR looks ready to merge, which I will plan to do tomorrow after confirmation from all reviewers no other changes required/requested.

tajila commented:
LGTM

llxia commented:
Thanks @XxAdi101xX for internal testing.
